src/utils/FileEmitter.ts

```diff
@@ -332,17 +330,24 @@ export class FileEmitter {
 		referenceId: string,
 		{ bundle, fileNamesBySource, outputOptions }: FileEmitterOutput
 	): void {
-		const fileName =
-			consumedFile.fileName ||
-			(typeof source === 'string' && fileNamesBySource.get(source)) ||
-			generateAssetFileName(consumedFile.name, source, outputOptions, bundle);
+		let fileName = consumedFile.fileName;
+
+		// Deduplicate assets if an explicit fileName is not provided
+		if (!fileName) {
+			const pregenerated = pregenerateAssetFileName(consumedFile.name, source, outputOptions);
+			// Reuse source hash if it was computed as part of the asset file name
+			const sourceHash = pregenerated.sourceHash ?? getSourceHash(source);
+			fileName = fileNamesBySource.get(sourceHash);
+			if (!fileName) {
+				fileName = makeUnique(pregenerated.fileName, bundle);
+				fileNamesBySource.set(sourceHash, fileName);
+			}
+		}
 
 		// We must not modify the original assets to avoid interaction between outputs
 		const assetWithFileName = { ...consumedFile, fileName, source };
 		this.filesByReferenceId.set(referenceId, assetWithFileName);
-		if (typeof source === 'string') {
-			fileNamesBySource.set(source, fileName);
-		}
```

Comment:
Removed these lines because `fileNamesBySource` is only being used in `finalizeAsset`, and the deduplication can be contained inside it.